src/core/server/http/base_path_proxy_server.ts

```diff
@@ -0,0 +1,162 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { RouteOptions, Server } from 'hapi-latest';
+import { Agent as HttpsAgent, ServerOptions as TlsOptions } from 'https';
+import { sample } from 'lodash';
+import { ByteSizeValue } from '../config/schema';
+import { DevConfig } from '../dev';
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+import { createServer, getServerOptions } from './http_tools';
+
+const alphabet = 'abcdefghijklmnopqrztuvwxyz'.split('');
+
+export interface BasePathProxyServerOptions {
+  httpConfig: HttpConfig;
+  devConfig: DevConfig;
+  isKibanaPath: (path: string) => boolean;
+  blockUntil?: () => Promise<void>;
+}
+
+export class BasePathProxyServer {
+  private server?: Server;
+  private httpsAgent?: HttpsAgent;
+
+  get basePath() {
+    return this.options.httpConfig.basePath;
+  }
+
+  get targetPort() {
+    return this.options.devConfig.basePathProxyTargetPort;
+  }
+
+  constructor(
+    private readonly log: Logger,
+    private readonly options: BasePathProxyServerOptions
+  ) {
+    const ONE_GIGABYTE = 1024 * 1024 * 1024;
+    options.httpConfig.maxPayload = new ByteSizeValue(ONE_GIGABYTE);
+
+    if (!options.httpConfig.basePath) {
+      options.httpConfig.basePath = `/${sample(alphabet, 3).join('')}`;
+    }
+  }
+
+  public async start() {
+    const { httpConfig, devConfig, blockUntil, isKibanaPath } = this.options;
+
+    const options = getServerOptions(httpConfig);
+    const server = createServer(options);
+
+    await server.register({ plugin: require('h2o2-latest') });
+
+    if (httpConfig.ssl.enabled) {
+      const tlsOptions = options.tls as TlsOptions;
+      this.httpsAgent = new HttpsAgent({
+        ca: tlsOptions.ca,
+        cert: tlsOptions.cert,
+        key: tlsOptions.key,
+        passphrase: tlsOptions.passphrase,
+        rejectUnauthorized: false,
+      });
+    }
+
+    // Always redirect from root URL to the URL with basepath.
+    server.route({
+      handler: (request, responseToolkit) => {
+        return responseToolkit.redirect(httpConfig.basePath);
+      },
+      method: 'GET',
+      path: '/',
+    });
+
+    // Before we proxy request to a target port we may want to wait until some
+    // condition is met (e.g. until target listener is ready).
+    let routeOptions: RouteOptions | undefined;
+    if (blockUntil !== undefined) {
+      routeOptions = {
+        pre: [
+          async (request, responseToolkit) => {
+            await blockUntil();
+            return responseToolkit.continue;
+          },
```

Comment:
This is interesting. So if we didn't have `blockUntil` abstraction, we'd hardcode waiting for the listener to be ready? (Just asking if I got that right) If that's true, then how does it look without this abstraction--would we need to reference `ClusterManager` here? My main motivation in asking is, wouldn't it always be a good idea to first wait for the listener on that port before proxying to it?

Reply:
> Do you mean we should make blockUntil as required option (not optional as it's now) and hence always register this pre hook?

Yes, either require blockUntil, or more generally, just go ahead and always run a default blockUntil that waits for a listener to be ready. Is that not possible because we don't have awareness of what listener to wait for?